### PR TITLE
Gatt dm generic serv

### PR DESCRIPTION
**Pull request by rakons**
Continuation of #288 as the previous fork loose the fork link after repository was made public.

### CODE REVIEW

File: include/bluetooth/common/gatt_dm.h [hunk 1]
```diff
@@ -252,11 +252,16 @@ const struct bt_gatt_attr *bt_gatt_dm_desc_next(
...
  * @param[in]     conn Connection object.
- * @param[in]     svc_uuid UUID of target service.
+ * @param[in]     svc_uuid UUID of target service
+ *                or NULL if any service should be discovered.
  * @param[in]     cb Callback structure.
  * @param[in,out] context Context argument that would be passed to
  *                callback functions
  *
+ * @note
+ * If @c svc_uuid is set to NULL all services may be discovered.
```
Comment (ru-fu) [inline include/bluetooth/common/gatt_dm.h]:
```suggestion
 * If @p svc_uuid is set to NULL, all services may be discovered.
```

File: include/bluetooth/common/gatt_dm.h [hunk 2]
```diff
@@ -252,11 +252,16 @@ const struct bt_gatt_attr *bt_gatt_dm_desc_next(
...
+ * To process next service see @ref bt_gatt_dm_continue function.
```
Comment (ru-fu) [inline include/bluetooth/common/gatt_dm.h]:
```suggestion
  * To process the next service, call @ref bt_gatt_dm_continue.
```

File: include/bluetooth/common/gatt_dm.h [hunk 3]
```diff
@@ -265,6 +270,20 @@ int bt_gatt_dm_start(struct bt_conn *conn,
 		     const struct bt_gatt_dm_cb *cb,
 		     void *context);
 
+/** @brief Continue discovery
```
Comment (ru-fu) [inline include/bluetooth/common/gatt_dm.h]:
```suggestion
/** @brief Continue service discovery.
```

File: include/bluetooth/common/gatt_dm.h [hunk 4]
```diff
@@ -265,6 +270,20 @@ int bt_gatt_dm_start(struct bt_conn *conn,
...
+ *
+ * This function continues service discovery.
+ * Call it after the previous data was released by @ref bt_gatt_dm_data_release.
+ *
+ * @param[in,out] dm Discovery Manager instance
```
Comment (ru-fu) [inline include/bluetooth/common/gatt_dm.h]:
```suggestion
 * @param[in,out] dm Discovery Manager instance.
```

File: include/bluetooth/common/gatt_dm.h [hunk 5]
```diff
@@ -252,11 +252,16 @@ const struct bt_gatt_attr *bt_gatt_dm_desc_next(
  * and call @ref bt_gatt_dm_data_release if it was successful.
  *
  * @param[in]     conn Connection object.
- * @param[in]     svc_uuid UUID of target service.
+ * @param[in]     svc_uuid UUID of target service
+ *                or NULL if any service should be discovered.
  * @param[in]     cb Callback structure.
  * @param[in,out] context Context argument that would be passed to
```
Comment (ru-fu) [inline include/bluetooth/common/gatt_dm.h]:
```suggestion
  * @param[in,out] context Context argument to be passed to
```

File: include/bluetooth/common/gatt_dm.h [hunk 6]
```diff
@@ -252,11 +252,16 @@ const struct bt_gatt_attr *bt_gatt_dm_desc_next(
...
  *                callback functions
```
Comment (ru-fu) [inline include/bluetooth/common/gatt_dm.h]:
```suggestion
  *                callback functions.
```

File: include/bluetooth/common/gatt_dm.h [hunk 7]
```diff
@@ -265,6 +270,20 @@ int bt_gatt_dm_start(struct bt_conn *conn,
 		     const struct bt_gatt_dm_cb *cb,
 		     void *context);
 
+/** @brief Continue discovery
+ *
+ * This function continues service discovery.
+ * Call it after the previous data was released by @ref bt_gatt_dm_data_release.
+ *
+ * @param[in,out] dm Discovery Manager instance
+ * @param[in]     context Context argument that would
```
Comment (ru-fu) [inline include/bluetooth/common/gatt_dm.h]:
```suggestion
 * @param[in]     context Context argument to
```

File: include/bluetooth/common/gatt_dm.h [hunk 8]
```diff
@@ -265,6 +270,20 @@ int bt_gatt_dm_start(struct bt_conn *conn,
...
+ *                be passed to callback functions
```
Comment (ru-fu) [inline include/bluetooth/common/gatt_dm.h]:
```suggestion
 *                be passed to callback functions.
```

File: include/bluetooth/common/gatt_dm.h [hunk 9]
```diff
@@ -265,6 +270,20 @@ int bt_gatt_dm_start(struct bt_conn *conn,
...
+ *
+ * @return 0 if the operation was successful.
```
Comment (ru-fu) [inline include/bluetooth/common/gatt_dm.h]:
```suggestion
 * @retval 0 If the operation was successful.
```

File: tests/gatt_dm/src/main.c [hunk 10]
```diff
@@ -309,16 +344,50 @@ void test_gatt_HIDS_chrc_by_uuid(void)
...
+	zassert_true(!bt_uuid_cmp(BT_UUID_DIS, serv_val->uuid), "Invalid service detected");
+	zassert_equal(5,
+		      bt_gatt_dm_attr_cnt(dm),
+		      "Unexpected number of attributes detected: %d",
+		      bt_gatt_dm_attr_cnt(dm));
+
+	dm = run_dm_next(dm);
+	zassert_is_null(dm, "Unexpected service detected");
+	/* ------------------------------------------------------ */
+	/* No cleanup here - cleanu up is done in run_dm_next */
```
Comment (ru-fu) [inline tests/gatt_dm/src/main.c]:
```suggestion
	/* No cleanup here - cleanup is done in run_dm_next */
```

**Comment by rakons** [comment]
@ru-fu - all requested changes implemented.

**Comment by rakons** [comment]
>>    One more thing: you can cross out one item from Limitations list that can be found in the gatt_dm.rst file:
    The discovery procedure can be targeted at the chosen service. Group type service discovery is not supported

> Please correct it before merging

Done

**Comment by kapi-no** [comment]
@pizi-nordic, Can you merge?

**Pull request merged**